### PR TITLE
RBAC CLI: list roles

### PR DESCRIPTION
**Pull request by peterschwarz**
This PR adds the command `splinter role list` to the CLI

# Testing

## Setup a local splinter node

These operations for setup can be done in one terminal.

First, initialize your splinter instance:

```
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path cli/Cargo.toml --features experimental -- cert generate
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path cli/Cargo.toml --features experimental -- database migrate
$ echo 02091a06cc46c5e0d88e89284936edb16800b03b56a8f1b7ec292f232b7c8385a2 > /tmp/splinter/data/allow_keys
```
(The last command adds the public key associated with the Cylinder JWT token we'll use in later steps.) 

Also, generate a key and add it to the allow_keys file:
```
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- keygen
$ cat ~/.cylinder/keys/$USER.pub >> /tmp/splinter/data/allow_keys
```

Start your splinter node:
```
SPLINTER_HOME=/tmp/splinter cargo run --manifest-path splinterd/Cargo.toml --features experimental -- --tls-insecure --enable-biome -vv --rest-api-endpoint http://127.0.0.1:8080
```

## List Roles

These operations can be done in a second terminal. (Note that the ouput has been formatted for readability, via `jq`) 

First, set the auth header value you will use when creating roles via curl commands:

```
export AUTH_TOKEN="Bearer Cylinder:eyJhbGciOiJzZWNwMjU2azEiLCJ0eXAiOiJjeWxpbmRlcitqd3QifQ==.eyJpc3MiOiIwMjA5MWEwNmNjNDZjNWUwZDg4ZTg5Mjg0OTM2ZWRiMTY4MDBiMDNiNTZhOGYxYjdlYzI5MmYyMzJiN2M4Mzg1YTIifQ==.tOMakxmebss0WGWcvKCQhYo2AAo3aaMDPS28y9nfVnMXiYq98Be08CdxB0gXCY5qYHZSw53+kjuIG+8gPhXLBA=="
```

List the current roles:

```
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- role list
ID  DISPLAY NAME
```

Add a role:
```
$  curl -s \
    -H "Authorization: $AUTH_TOKEN" \
    -H "Content-Type: application/json" \
    -X POST \
    -d '{"role_id": "status_read", "display_name": "Status Reader", "permissions": ["status.read"]}' \
     localhost:8080/authorization/roles
```

Now, list the roles again:

```
$ cargo run --manifest-path cli/Cargo.toml --features experimental -- role list
ID          DISPLAY NAME
status_read Status Reader
```

Likewise, the output can be rendered as csv with the `--format csv` argument.

### CODE REVIEW

File: cli/man/splinter-role.1.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+% SPLINTER-ROLE(1) Cargill, Incorporated | Splinter Commands
+<!--
+  Copyright 2018-2020 Cargill Incorporated
```
Comment (agunde406) [inline cli/man/splinter-role.1.md:3]: 2020 -> 2021

**Review by agunde406 — CHANGES_REQUESTED** [review]
Tested.

File: cli/man/splinter-role.1.md (hunk 1 above)
Comment (peterschwarz) [reply cli/man/splinter-role.1.md]: fixed

**Review by agunde406 — DISMISSED** [review]

**Review by agunde406 — APPROVED** [review]

File: cli/src/action/api/rbac.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    }
+
+    fn load_page(base_url: &str, auth: &str, link: &str) -> Option<Result<RoleList, CliError>> {
+        let result = Client::new()
+            .get(&format!("{}{}", base_url, link))
+            .header("SplinterProtocolVersion", RBAC_PROTOCOL_VERSION)
+            .header("Authorization", auth)
+            .send()
+            .map_err(|err| CliError::ActionError(format!("Failed to fetch node ID: {}", err)))
```
Comment (ltseeley) [inline cli/src/action/api/rbac.rs:69]: s/node ID/roles/

File: cli/src/action/api/rbac.rs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            });
+
+        Some(result)
+    }
+}
+
+impl<'a> Iterator for RoleIter<'a> {
+    type Item = Result<Role, CliError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
```
Comment (ltseeley) [inline cli/src/action/api/rbac.rs:104]: This function is hard to understand (at least for my morning brain). Is there a different way to write it that's easier to understand? If not, please add some more comments here and/or to the `current_page` field.

File: cli/man/splinter-role-list.1.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+**splinter role list** \[**FLAGS**\] \[**OPTIONS**\]
+
+DESCRIPTION
+===========
+This command lists all of the roles the local node has available. This command
+displays abbreviated information pertaining to roles in columns, with the
+headers `ID` and `DISPLAY NAME`. This makes it possible to verify that
+roles have been successfully created as well as being able to access the
+available role ID for use when assigning a role to an identity. The information
+displayed will be is only relevant to the queried splinter nodes.
```
Comment (ltseeley) [inline cli/man/splinter-role-list.1.md:24]: s/ will be// and probably s/nodes/node/

File: cli/man/splinter-role.1.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+SYNOPSIS
+========
+
+**splinter** **role** \[**FLAGS**\] \[**SUBCOMMAND**\]
+
+DESCRIPTION
+===========
+
+This command provides subcommands for checking and modifying the maintenance
+status of the Splinter daemon.
```
Comment (ltseeley) [inline cli/man/splinter-role.1.md:23]: Copypasta

Comment (peterschwarz) [reply cli/man/splinter-role.1.md]: Fixed

File: cli/man/splinter-role-list.1.md (hunk 4 above)
Comment (peterschwarz) [reply cli/man/splinter-role-list.1.md]: Fixed

File: cli/src/action/api/rbac.rs (hunk 3 above)
Comment (peterschwarz) [reply cli/src/action/api/rbac.rs:104]: Added extensive comments in the method.

File: cli/src/action/api/rbac.rs (hunk 2 above)
Comment (peterschwarz) [reply cli/src/action/api/rbac.rs]: Fixed

File: cli/src/action/api/rbac.rs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        }
+    }
+
+    fn load_page(base_url: &str, auth: &str, link: &str) -> Option<Result<RoleList, CliError>> {
+        let result = Client::new()
+            .get(&format!("{}{}", base_url, link))
+            .header("SplinterProtocolVersion", RBAC_PROTOCOL_VERSION)
+            .header("Authorization", auth)
+            .send()
+            .map_err(|err| CliError::ActionError(format!("Failed to role list page: {}", err)))
```
Comment (ltseeley) [inline cli/src/action/api/rbac.rs:69]: Should this be "Failed to get..."?

Comment (peterschwarz) [reply cli/src/action/api/rbac.rs]: fixed

**Review by ltseeley — APPROVED** [review]

**Review by agunde406 — APPROVED** [review]

**Pull request merged**